pkg/controller/tenant/tenant_controller.go

```diff
@@ -0,0 +1,198 @@
+	defer utilruntime.HandleCrash()
+	defer tc.queue.ShutDown()
+
+	klog.Infof("Starting tenant controller.")
+	defer klog.Infof("Shutting down tenant controller")
+
+	if !controller.WaitForCacheSync("tenant", stopCh, tc.listerSynced) {
+		return
+	}
+
+	klog.V(5).Info("Starting workers of tenant controller")
+	for i := 0; i < workers; i++ {
+		go wait.Until(tc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// worker processes the queue of tenant objects.
+func (tc *tenantController) worker() {
+	workFunc := func() bool {
+		key, quit := tc.queue.Get()
+		if quit {
+			return true
+		}
+		defer tc.queue.Done(key)
+
+		err := tc.processQueue(key.(string))
+		if err == nil {
+			// no error, forget this entry and return
+			tc.queue.Forget(key)
+			return false
+		} else {
+			// rather than wait for a full resync, re-add the tenant to the queue to be processed
+			tc.queue.AddRateLimited(key)
+			utilruntime.HandleError(err)
+		}
+		return false
+	}
+
+	for {
+		quit := workFunc()
+
+		if quit {
+			return
+		}
+	}
+}
+
+// enqueue adds an object to the controller work queue
+func (tc *tenantController) enqueue(obj interface{}) {
+	klog.Infof("Starting tenant enque.")
+
+	tenant, ok := obj.(*v1.Tenant)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("Not a tenant object: %v", obj))
+		return
+	}
+
+	tc.queue.Add(tenant.Name)
+}
+
+// processQueue looks for a tenant with the specified name and synchronizes it
+func (tc *tenantController) processQueue(tenantName string) (err error) {
+	klog.Infof("Starting processsing queue for tenant: %v", tenantName)
+
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing tenant %q (%v)", tenantName, time.Since(startTime))
+	}()
+
+	_, err = tc.lister.Get(tenantName)
+	if errors.IsNotFound(err) {
+		klog.Infof("tenant has been deleted %v", tenantName)
+		return nil
+	}
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Unable to retrieve tenant %v from store: %v", tenantName, err))
+		return err
+	}
+
+	return tc.syncHandler(tenantName)
+}
+
+// syncTenant creates the default resources for a new tenant
+func (tc *tenantController) syncTenant(tenantName string) (err error) {
+	klog.Infof("Starting tenant initialization: %v", tenantName)
+
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished initializing tenant %q (%v)", tenantName, time.Since(startTime))
+	}()
+
+	createFailures := []error{}
+
+	// Create Default namespaces
+	for _, nsName := range tenant_defult_namespaces {
+
+		ns := v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{Tenant: tenantName, Name: nsName},
+		}
+		if _, err := tc.kubeClient.CoreV1().NamespacesWithMultiTenancy(tenantName).Create(&ns); err != nil && !errors.IsAlreadyExists(err) {
+			createFailures = append(createFailures, err)
+		}
+	}
+
+	ret := utilerrors.Flatten(utilerrors.NewAggregate(createFailures))
+	klog.Infof("Tenant initialization result of %v : %v", tenantName, ret)
```

Comment:
The info here might be a bit confusing. When the initialization is successful, ret will be nil, the message will read" result: nil", which might make the log reader think sth is wrong.

Reply:
Thank you for capturing this. Changed. Will return nil if the initialization is successful.